crates/storage/provider/src/providers/mod.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{BlockHashProvider, BlockProvider, Error, HeaderProvider, StateProviderFactory};
+use reth_db::{
+    database::{Database, DatabaseGAT},
+    tables,
+    transaction::DbTx,
+};
+use reth_interfaces::Result;
+use reth_primitives::{rpc::BlockId, Block, BlockHash, BlockNumber, ChainInfo, Header, H256, U256};
+use std::sync::Arc;
+
+mod historical;
+pub use historical::{HistoricalStateProvider, HistoricalStateProviderRef};
+
+mod latest;
+pub use latest::{LatestStateProvider, LatestStateProviderRef};
+
+/// A common provider that fetches data from a database.
+///
+/// This provider implements most provider or provider factory traits.
+pub struct ShareableDatabase<DB: Database> {
```

Comment:
Perhaps we should hide this as an implementation detail? I.e. mark `Database: Clone` and then have this part of the MDBX DB impl

Reply:
yeah, that's kinda what @mattsse and me were talking about above